icu4c/source/common/unicode/char16ptr.h

```diff
@@ -309,13 +310,19 @@ inline OldUChar *toOldUCharPtr(char16_t *p) {
 
 #ifndef U_FORCE_HIDE_INTERNAL_API
 /**
- * Is T convertible to a std::u16string_view or to a 16-bit std::wstring_view?
+ * Is T convertible to a std::u16string_view or some other 16-bit string view?
  * @internal
  */
 template<typename T>
 constexpr bool ConvertibleToU16StringView =
-    std::is_convertible_v<T, std::u16string_view> ||
-    (U_SIZEOF_WCHAR_T==2 && std::is_convertible_v<T, std::wstring_view>);
+    std::is_convertible_v<T, std::u16string_view>
+#if !U_CHAR16_IS_TYPEDEF
+    || std::is_convertible_v<T, std::basic_string_view<uint16_t>>
```

Comment:
I do not think that we should support `std::basic_string_view<uint16_t>`; uint16_t does not have a standard std::char_traits specification, so I am not sure how that will work (and indeed, how exactly it works could be changed by someone defining their own specialization), and it obviously is incompatible with string literals.

We support string views because we want to be compatible with idiomatic modern code. `std::basic_string_view<uint16_t>` may be modern, but idiomatic it is not.

Of course everything _exists_,  but compare the number of results for https://github.com/search?q=%22basic_string_view%3Cuint16_t%3E%22&type=code and https://github.com/search?q=%22std%3A%3Au16string_view%22&type=code or https://github.com/search?q=%22std%3A%3Awstring_view%22&type=code.

Reply:
FWIW I didn't think that supporting uint16_t, given ICU's and its users' history, was a bad idea. (We know that some people still use UChar=uint16_t, although UChar is nearly gone from ICU C++).

But if uint16_t as a string unit type is not fully supported by the STL, then that could be more frustrating than useful.

Reply:
What exactly does "not fully supported" mean here? Is it something that could actually effect the operations that are done here, which are nothing more than calling the `data()` and `size()` members?

Reply:
According to my reading https://eel.is/c++draft/string.view.template#general-1, `std::basic_string_view<uint16_t>` should be ill-formed unless `char_traits<uint16_t>` contains what is needed, which the standard does not mandate; so depending on the whims of the various bits of the standard library implementation, there is no guarantee that this class compiles.

Reply:
Hm, does that mean that even the two mentions of `std::basic_string_view<uint16_t>` as a possible type here would fail to compile, or just that any code that tried to actually use that could fail to compile with some standard library implementations?

Our CI tests this with the GNU, LLVM and MSVC standard library implementations and at least with those the mentions of it as a possible type here compiles just fine.

So to gauge whether this really is a useful thing to support, I now took a closer look at what Clang 18 actually does with the following piece of example code together with the implementation currently proposed in this PR:

```
int test() {
  static constexpr UChar TEXT[] = { 0x66, 0x6f, 0x6f, 0x00 };
  return do_something(icu::UnicodeString(TEXT));
}
```

Both with the default `UChar` = `char16_t` and also with `-DUCHAR_TYPE=uint16_t` this really did result in the compiler directly creating a string view object of known size, which is what I had assumed that it would do (before you pointed out that this data type wasn't standard), so as long as we keep having both `UChar` and `UCHAR_TYPE` in the public API, I think it's very nice that this just works (and that switching the data type doesn't suddenly and unexpectedly change the runtime cost from fixed to linear).

So unless there's any real reason to believe that this would make it impossible to compile ICU4C itself with some standard library implementation that it otherwise would be reasonable to expect to be used, I'd say that supporting this adds some real actual value.

Reply:
> So unless there's any real reason to believe that this would make it impossible to compile ICU4C itself with some standard library implementation that it otherwise would be reasonable to expect to be used, I'd say that supporting this adds some real actual value.

wfm, ok to keep the uint16_t support until we learn otherwise
and I think we can consider this part "draft" and change it if forced by the standard, or maybe with compiler ifdefs

Reply:
> Our CI tests this with the GNU, LLVM and MSVC standard library implementations and at least with those the mentions of it as a possible type here compiles just fine.

It turns out that our CI also tests with one more implementation, the CIFuzz tests, which are currently not run as presubmit, only as postsubmit (PR #3142), use some other standard library implementation than the others and that one doesn't define `std::char_traits<uint16_t>` and that makes the compilation fail spectacularly (no such thing as `std::is_convertible_v<>` failing silently when the type compared to isn't defined).

I failed to find what standard library implementation CIFuzz actually uses (it's a layer cake of Docker images, where I quickly lost track) but the error messages look exactly like error messages one gets from the LLVM release 19 libc++ so that seems like a really good guess.

I've now sent PR #3144 for review to simply disable the use of `std::basic_string_view<uint16_t>` when using a too new libc++. That should be easy to find and delete in the future, once the tested-for version is too old to be relevant anymore.

Reply:
(Late to the party-- just back from vacation)  I feel obligated to point out that Apple _does_ still define `UChar` as `uint16_t`, as we have too many internal clients that are assuming that and I'm not sure I can get them all to change.  But unless Robin's issue is causing real problems in practice, I'm fine with what Fredrik is proposing here.

Reply:
> I feel obligated to point out that Apple _does_ still define `UChar` as `uint16_t`,

The changes made should not break anything for Apple (or anyone else) and the macOS tests on our CI still pass as does testing locally on my own Mac. But it'd be most valuable if you could verify that everything still works as expected for all your internal use of ICU4C at Apple (whatever that might be) at commit 7cad791fb666e6759bd7aaea0461eebafbda9cfe or later.

Reply:
I'll be testing this pretty soon, although might not really know if we're in trouble until after ICU 76 goes out the door.  If there are problems, you'll definitely hear from me in the next couple months.